### PR TITLE
Improve jdk download retries

### LINKED ISSUE #40531: Default to the bundled JDK for integration tests in CI
Right now we only test with the bundled JDK version in the packaging tests but for other integration tests  when `RUNTIME_JAVA_HOME` is not set.
Since we always set the lather in CI we don't get any coverage with respect to integration tests for the bundled JDK there. 

We should move to default to the bundled JDK in CI for everything other than the specific java matrix testing. Since the integ test distribution does not bundle the jdk ( and we don't want it to ) we need a way in the build to provision and use the same jdk as the one we bundle for those tests.

### PR DESCRIPTION
**Pull request by alpar-t**
Look like the `--retry` argument is not always effective.

Closes #40531

### CODE REVIEW

File: distribution/docker/src/docker/Dockerfile [hunk 1]
```diff
@@ -16,9 +16,9 @@ FROM centos:7 AS builder
 ENV PATH /usr/share/elasticsearch/bin:$PATH
 ENV JAVA_HOME /opt/jdk-${jdkVersion}
 
-RUN curl -L --retry 8 -s -S ${jdkUrl} --continue-at - --output /tmp/jdk.tar.gz && \
-    tar -C /opt -zxf /tmp/jdk.tar.gz && \
-    rm -Rf /tmp/jdk.tar.gz
+RUN for iter in {1..10}; do curl -L -s -S ${jdkUrl} | tar -C /opt -zx &&\
```
Comment (dliappis) [inline distribution/docker/src/docker/Dockerfile]: nit: let's have a space between `&&` and `\` for consistency i.e. `&& \`

File: distribution/docker/src/docker/Dockerfile [hunk 2]
```diff
@@ -16,9 +16,9 @@ FROM centos:7 AS builder
...
+    exit_code=0 && break || exit_code=\$? && echo "download error: retry \$iter in 10s" && sleep 10; done;
+
```
Comment (dliappis) [inline distribution/docker/src/docker/Dockerfile]: One new line is enough, I think we can remove this.

**Review by dliappis — CHANGES_REQUESTED** [review]
Thanks! I left a few comments, mainly highlighting the need to emit the error at the end.

**Review by mark-vieira — APPROVED** [review]
LGTM I suppose. Pretty annoying that `--retry` doesn't do what we want. Did we discover what kind of errors we are encountering that aren't caught by `--retry`?

**Comment by dliappis** [comment]
> Did we discover what kind of errors we are encountering that aren't caught by `--retry`?

Connection resets -- which is something that occasionally happens from the GitHub URL we are relying -- aren't being retried (others reporting the same [here](https://superuser.com/questions/1083589/curl-56-recv-failure-connection-reset-by-peer)).

For the record `curl` already covers connection refused (via https://github.com/curl/curl/issues/1036) and will also retry connection timeouts (if [--connect-timeout](https://github.com/curl/curl/issues/4461) has been set).

**Comment by atorok** [comment]
but `--connection-timeout`is not available in the curl version we use here.

File: distribution/docker/src/docker/Dockerfile [hunk 3]
```diff
@@ -16,10 +16,10 @@ FROM centos:7 AS builder
...
 ENV JAVA_HOME /opt/jdk-${jdkVersion}
 
-RUN curl -L --retry 8 -s -S ${jdkUrl} --continue-at - --output /tmp/jdk.tar.gz && \
-    tar -C /opt -zxf /tmp/jdk.tar.gz && \
-    rm -Rf /tmp/jdk.tar.gz
-
+RUN for iter in {1..10}; do curl -L -s -S ${jdkUrl} | tar -C /opt -zx && \
+    exit_code=0 && break || exit_code=\$? && echo "download error: retry $iter in 10s" && sleep 10; done; \
+    (exit $exit_code)
+
```
Comment (dliappis) [inline distribution/docker/src/docker/Dockerfile]: (super) nit: we could remove the 4 space characters in line 22 (so that there's just a newline).

**Review by dliappis — APPROVED** [review]
LGTM. I left a minor comment, there's no need for another review cycle.

**Pull request merged**